base_exception/models/base_exception.py

```diff
@@ -0,0 +1,251 @@
+        object that inherits for base.exception.
+        for sale :
+        @api.constrains('ignore_exception',)
+        def sale_check_exception(self):
+            ...
+            ...
+            self._check_exception
+        """
+        exception_ids = self.detect_exceptions()
+        if exception_ids:
+            exceptions = self.env['exception.rule'].browse(exception_ids)
+            raise ValidationError('\n'.join(exceptions.mapped('name')))
+
+    @api.multi
+    def test_exceptions(self):
+        """
+        Condition method for the workflow from draft to confirm
+        """
+        if self.detect_exceptions():
+            return False
+        return True
+
+    @api.multi
+    def detect_exceptions(self):
+        """returns the list of exception_ids for all the considered base.exceptions
+        """
+        if not self:
+            return []
+        exception_obj = self.env['exception.rule']
+        all_exceptions = exception_obj.sudo().search(
+            [('rule_group', '=', self[0].rule_group)])
+        model_exceptions = all_exceptions.filtered(
+            lambda ex: ex.model == self._name)
+        sub_exceptions = all_exceptions.filtered(
+            lambda ex: ex.model != self._name)
+
+        all_exception_ids = []
+        for obj in self:
+            if obj.ignore_exception:
+                continue
+            exception_ids = obj._detect_exceptions(
+                model_exceptions, sub_exceptions)
+            obj.exception_ids = [(6, 0, exception_ids)]
+            all_exception_ids += exception_ids
+        return all_exception_ids
+
+    @api.model
+    def _exception_rule_eval_context(self, obj_name, rec):
+        return {obj_name: rec,
+                'self': self.env[rec._name],
+                'object': rec,
+                'obj': rec,
+                'env': self.env,
+                'cr': self.env.cr,
+                'uid': self.env.user.id,
+                'user': self.env.user,
+                'time': time,
+                # copy context to prevent side-effects of eval
+                'context': self._context.copy()}
+
+    @api.model
+    def _rule_eval(self, rule, obj_name, rec):
+        expr = rule.code
+        space = self._exception_rule_eval_context(obj_name, rec)
+        try:
+            safe_eval(expr,
+                      space,
+                      mode='exec',
+                      nocopy=True)  # nocopy allows to return 'result'
+        except Exception as e:
+            raise UserError(
+                _('Error when evaluating the exception.rule '
+                  'rule:\n %s \n(%s)') % (rule.name, e))
+        return space.get('failed', False)
+
+    @api.multi
+    def _detect_exceptions(self, model_exceptions,
+                           sub_exceptions):
+        self.ensure_one()
+        exception_ids = []
+        next_state_rule = False
+        for rule in model_exceptions:
+            if self._rule_eval(rule, self.rule_group, self):
```

Comment:
@mourad-ehm IMHO Same as above, you could decorate your _rule_eval function with api.multi and remove both self.rule_group and self arguments.

Reply:
I can't do this because I use the same method for object and lines. 
code: if self._rule_eval(rule, group_line, obj_line):

Reply:
ping @rousseldenis

Reply:
ok!

Reply:
You can approve :)

Reply:
oups. For sure